File: solutions/java/squeaky-clean/1/src/main/java/SqueakyClean.java

class SqueakyClean {
    static String clean(String identifier) {
        StringBuilder cleaner = new StringBuilder();

        boolean camel = false;

        for (int i = 0; i < identifier.length(); i++) {
            char character = identifier.charAt(i);

            if (Character.isISOControl(character)) {
                cleaner.append("CTRL");
            } else if (Character.isWhitespace(character)) {
                cleaner.append("_");
            } else if (character == '-') {
                camel = true;
            } else if (SqueakyClean.isValidLetter(character)) {
                if (camel) {
                    cleaner.append(Character.toUpperCase(character));
                    camel = false;
                } else {
                    cleaner.append(character);
                }
            }
        }

        return cleaner.toString();
    }

    static boolean isValidLetter(char letter) {
        return Character.isLetter(letter) && ('ω' < letter || letter < 'α');
    }
}

File: solutions/csharp/squeaky-clean/1/SqueakyClean.cs
using System.Text;

public static class Identifier
{
    public static string Clean(string identifier)
    {
        var s = new StringBuilder();

        var camelActive = false;
        
        foreach (var c in identifier)
        {

            switch (c)
            {
                case ' ':
                    s.Append('_');
                    continue;
                case '\0':
                    s.Append("CTRL");
                    continue;
                case '-':
                    camelActive = true;
                    continue;
            }

            if (camelActive)
            {
                s.Append(char.ToUpper(c));
                camelActive = false;
                continue;
            }

            if (!char.IsLetter(c) || char.IsBetween(c, 'α', 'ω'))
            {
                continue;
            }
            
            s.Append(c);
        }
        
        return s.ToString();
    }
}
